src/java/org/apache/cassandra/index/sai/disk/v1/MemtableIndexWriter.java

```diff
@@ -0,0 +1,177 @@
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.disk.v1;
+
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.base.Stopwatch;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.carrotsearch.hppc.LongArrayList;
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.marshal.AbstractType;
+import org.apache.cassandra.db.rows.Row;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.disk.MemtableTermsIterator;
+import org.apache.cassandra.index.sai.disk.PerIndexWriter;
+import org.apache.cassandra.index.sai.disk.format.IndexComponent;
+import org.apache.cassandra.index.sai.disk.format.IndexDescriptor;
+import org.apache.cassandra.index.sai.disk.v1.trie.InvertedIndexWriter;
+import org.apache.cassandra.index.sai.memory.MemtableIndex;
+import org.apache.cassandra.index.sai.memory.RowMapping;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.index.sai.utils.TypeUtil;
+import org.apache.cassandra.utils.Pair;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+
+/**
+ * Column index writer that flushes indexed data directly from the corresponding Memtable index, without buffering index
+ * data in memory.
+ */
+public class MemtableIndexWriter implements PerIndexWriter
+{
+    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private final IndexDescriptor indexDescriptor;
+    private final IndexContext indexContext;
+    private final MemtableIndex memtable;
+    private final RowMapping rowMapping;
+
+    public MemtableIndexWriter(MemtableIndex memtable,
+                               IndexDescriptor indexDescriptor,
+                               IndexContext indexContext,
+                               RowMapping rowMapping)
+    {
+        assert rowMapping != null && rowMapping != RowMapping.DUMMY : "Row mapping must exist during FLUSH.";
+
+        this.indexDescriptor = indexDescriptor;
+        this.indexContext = indexContext;
+        this.memtable = memtable;
+        this.rowMapping = rowMapping;
+    }
+
+    @Override
+    public void addRow(PrimaryKey key, Row row, long sstableRowId)
+    {
+        // Memtable indexes are flushed directly to disk with the aid of a mapping between primary
+        // keys and row IDs in the flushing SSTable. This writer, therefore, does nothing in
+        // response to the flushing of individual rows.
+    }
+
+    @Override
+    public void abort(Throwable cause)
+    {
+        logger.warn(indexContext.logMessage("Aborting index memtable flush for {}..."), indexDescriptor.descriptor, cause);
+        indexDescriptor.deleteColumnIndex(indexContext);
+    }
+
+    @Override
+    public void complete(Stopwatch stopwatch) throws IOException
```

Comment:
nit: I wonder if it would be nicer to have the `RowMapping` just passed here rather than passed to the constructor and modified in another object before our own no-op `addRow()` calls. We can't completely encapsulate it, because there may be multiple column indexes.

It might also be a good idea to `assert` at the start of `complete()` here that the mapping is already complete.

Reply:
I agree with asserting that the `RowMapping` is complete but I don't think we would want to pass it in `complete` because this forms part of the `PerIndexWriter` and is used by classes that have no knowledge of the `RowMapping`.

Reply:
Yeah, that's fair